### PR TITLE
Implement graph-based representation of params

### PR DESCRIPTION
**Pull request by zuevmaxim**

### CODE REVIEW

File: autotm/algorithms_for_tuning/genetic_algorithm/ga.py [hunk 1]
```diff
@@ -817,106 +497,39 @@ def run(self, verbose=False, visualize_results=False) -> Individual:
...
-                    population = estimate_fitness(population)
-                    self.save_params(population)
-            elif self.calc_scheme == "type2" and self.surrogate:
-                population = self._calculate_uncertain_res(population, iteration_num=ii)
-                self.save_params(population)
-
-            ###
+            population = self.run_fitness(population, surrogate_iteration, ii)
+
+            # TODO (pipeline) Mutations collection is disabled
```
Comment (fonhorst) [inline autotm/algorithms_for_tuning/genetic_algorithm/ga.py:515]: Are these comments lines necessary?

File: autotm/utils.py [hunk 2]
```diff
@@ -105,13 +104,14 @@ def parallelize_dataframe(df: pd.DataFrame, func, n_cores, return_type="df", **k
...
-    filename: str = "/var/log/tm-alg.txt", uid: Optional[str] = None
+        filename: str = "/var/log/tm-alg.txt", uid: Optional[str] = None,
+        quiet: bool = False
 ) -> Dict[str, Any]:
     if filename is not None:
         if uid:
             dirname = os.path.dirname(filename)
             file, ext = os.path.splitext(os.path.basename(filename))
-            log_filename = os.path.join(dirname, f"{file}-{uid}.{ext}")
+            log_filename = os.path.join(dirname, f"{file}-{uid}{ext}")
```
Comment (fonhorst) [inline autotm/utils.py:114]: why removing the dot before the extension?

File: distributed/kube_fitness/kube_fitness/tm.py [hunk 3]
```diff
@@ -336,8 +336,9 @@ def fit_tm_of_individual(dataset: str,
             }
         except SoftTimeLimitExceeded as ex:
             raise ex
-        except Exception:
-            logger.exception(msg="Fitness calculation problem")
+        except Exception as e:
+            logger.error("Fitness calculation problem")
```
Comment (fonhorst) [inline distributed/kube_fitness/kube_fitness/tm.py:340]: logger.error has 'exc_info'.

if 'exc_info' is True, the logger will add info about the current exception.

So, no need for separate 'logger.exception' call.

File: examples/experiments/experiment.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    with open(file_name, 'a') as logfile:
+        for dataset_name in datasets:
+            for use_pipeline in [False, True]:
+                for _ in range(10):
+                    start_time = time.time()
+                    used_fitness, best_fitness = suppress_stdout(lambda: run_single_experiment(dataset_name, date, use_pipeline))
+                    elapsed_time = time.time() - start_time
+                    elapsed_minutes, elapsed_seconds = divmod(int(elapsed_time), 60)
+                    stats = ", ".join(f"({i}, {fi})" for i, fi in zip(used_fitness, best_fitness))
+                    print(f"Time: {elapsed_minutes} min {elapsed_seconds} sec, use_pipeline: {use_pipeline}, dataset: {dataset_name}, (i, f(i)): {stats}")
```
Comment (fonhorst) [inline examples/experiments/experiment.py:101]: may be replace with logger

File: examples/experiments/experiment.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    git_hash = get_git_hash()
+    print(f'Git hash: {git_hash}')
+
+    date = datetime.datetime.now().strftime('%y%m%d-%H%M%S')
+    file_name = f"log-{date}.txt"
+    with open(file_name, 'a') as logfile:
+        for dataset_name in datasets:
+            for use_pipeline in [False, True]:
+                for _ in range(10):
+                    start_time = time.time()
```
Comment (fonhorst) [inline examples/experiments/experiment.py:96]: may be replaced with a time measuring context manager

**Review by fonhorst — APPROVED** [review]
Minor changes is needed, though nothing really important. I will merge.

**Comment by fonhorst** [comment]
Manually merged after conflicts solving.

See commit 2396923289530289ee4dba91e1dec20f7f442ad3

File: autotm/algorithms_for_tuning/genetic_algorithm/ga.py [hunk 6]
```diff
@@ -445,8 +229,11 @@ def _calculate_uncertain_res(self, generation, iteration_num: int, proc=0.3):
         return calculated
 
     def save_params(self, population):
+        if any(not isinstance(ind.dto.params, FixedListParams) for ind in population):
+            # TODO embeddings
```
[the remark this answers is not in the document]
Comment (zuevmaxim) [reply autotm/algorithms_for_tuning/genetic_algorithm/ga.py:233]: Fixed in #23

File: autotm/utils.py (hunk 2 above)
Comment (zuevmaxim) [reply autotm/utils.py:114]: Because `ext` already includes a dot

File: examples/experiments/experiment.py (hunk 4 above)
Comment (zuevmaxim) [reply examples/experiments/experiment.py:101]: This is an experimental script, it is used only in experiments

File: examples/experiments/experiment.py (hunk 5 above)
Comment (zuevmaxim) [reply examples/experiments/experiment.py:96]: Could you please elaborate? However, I don't this it is vital, as this is used only in experiments

**Pull request closed without merging**